tests/test_nsjail.py

```diff
@@ -184,31 +200,27 @@ def test_log_parser(self):
             log.output
         )
 
-    def test_shm_and_tmp_not_mounted(self):
-        for path in ("/dev/shm", "/run/shm", "/tmp"):
-            with self.subTest(path=path):
-                code = dedent(f"""
-                    with open('{path}/test', 'wb') as file:
-                        file.write(bytes([255]))
-                """).strip()
+    def test_tmp_not_mounted(self):
+        code = dedent("""
+            with open('/tmp/test', 'wb') as file:
+                file.write(bytes([255]))
+        """).strip()
 
-                result = self.nsjail.python3(code)
-                self.assertEqual(result.returncode, 1)
-                self.assertIn("No such file or directory", result.stdout)
-                self.assertEqual(result.stderr, None)
+        result = self.nsjail.python3(code)
+        self.assertEqual(result.returncode, 1)
+        self.assertIn("No such file or directory", result.stdout)
+        self.assertEqual(result.stderr, None)
 
-    def test_multiprocessing_shared_memory_disabled(self):
+    def test_multiprocessing_shared_memory_limited(self):
         code = dedent("""
-            from multiprocessing.shared_memory import SharedMemory
-            try:
-                SharedMemory('test', create=True, size=16)
-            except FileExistsError:
-                pass
+            from multiprocessing import shared_memory
+            shm_a = shared_memory.SharedMemory(create=True, size=45_000_000)
+            shm_a.buf[:45_000_000] = bytearray([1] * 45_000_000)
         """).strip()
 
         result = self.nsjail.python3(code)
         self.assertEqual(result.returncode, 1)
-        self.assertIn("Function not implemented", result.stdout)
+        self.assertIn("File too large", result.stdout)
```

Comment:
The file size limit is a different thing to the tmpfs size, and I think this is just testing the file size limit (`rlimit_fsize` parameter in nsjail, set to 1mb by default). To test the tmpfs size we'd have to create a bunch of files below 1mb. We should then get a different error, `OSError: [Errno 28] No space left on device`

Reply:
I've picked up this PR again, and in my testing, the process crashes with code 135 (`SIGBUS`). To trigger this, the code must write to the shared memory buffer. Merely creating the shared memory object is not enough.

What I've been stuck on is understanding what's going on when just the shared memory object is created. [Here](https://github.com/python/cpython/blob/main/Lib/multiprocessing/shared_memory.py#L104-L115) is the CPython code for it.

First it calls `shm_open`, and it has the `O_CREAT` flag set since `create=True` is passed in this test. This initially creates an empty file as documented [here](https://www.man7.org/linux/man-pages/man3/shm_open.3.html):

> A new shared memory object initially has zero length—the size of the object can be set using [ftruncate(2)](https://www.man7.org/linux/man-pages/man2/ftruncate.2.html).  The newly allocated bytes of a shared memory object are automatically initialized to 0.

Indeed, Python afterwards calls `ftruncate` with the size. This will set the file to exactly the given size (in this case, it has to extend the file from a size of 0 to the given size).

> The truncate() and ftruncate() functions cause the regular file named by path or referenced by fd to be truncated to a size of precisely length bytes.

Indeed, when I `stat` the file, I can see `st_size` is set to the expected size. Finally, it `mmap`s the entire file.

The strange thing is that despite the cgroup memory limit being in effect, it's possible to create files in `/dev/shm` much larger than the memory limit. For example, with a memory limit of `52428800`, I can create a file 13 times that size:

```
python -m snekbox 'from multiprocessing.shared_memory import SharedMemory; shm = SharedMemory(create=True, size=52428800 * 13); import os; print(os.stat("/dev/shm/" + shm.name))' --time_limit 0

os.stat_result(st_mode=33152, st_ino=2, st_dev=60, st_nlink=1, st_uid=65534, st_gid=65534, st_size=681574400, st_atime=1693161926, st_mtime=1693161926, st_ctime=1693161926)
```

However, when I inspect the memory in the cgroup (while the jailed process is running with a sleep), it is much smaller than the size of the file.

```
cat /sys/fs/cgroup/NSJAIL.92/memory.current 
14569472
```
```
cat /sys/fs/cgroup/NSJAIL.92/memory.stat | grep shm
shmem 0
shmem_thp 0
```
```
cat /sys/fs/cgroup/NSJAIL.100/memory.stat | grep map
file_mapped 0
```

The [docs](https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst#n1397) state that `shmem` in `memory.stat` should include shared memory and anything backed by tmpfs:

> shmem
> Amount of cached filesystem data that is swap-backed, such as tmpfs, shm segments, shared anonymous mmap()s

It seems that the cgroup memory controller is unable to capture memory usage from shared memory, which is a concern.

If the shared memory size is too large (unclear how large is too large), at least the `mmap` will fail:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.11/multiprocessing/shared_memory.py", line 115, in __init__
    self._mmap = mmap.mmap(self._fd, size)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 12] Cannot allocate memory
```

Reply:
It seems like no memory is consumed by just using `ftruncate`:

```
python -m snekbox '
import os
import _posixshmem
try:
   fd = _posixshmem.shm_open("test", os.O_CREAT | os.O_EXCL | os.O_RDWR, mode=0o600)
   os.ftruncate(fd, 52428800 * 5000000)
   print(os.stat(fd))
finally:
   _posixshmem.shm_unlink("test")
' --time_limit 0
```

```
os.stat_result(st_mode=33152, st_ino=2, st_dev=60, st_nlink=1, st_uid=65534, st_gid=65534, st_size=262144000000000, st_atime=1693165871, st_mtime=1693165871, st_ctime=1693165871)
```

That's 262 TB which would definitely run out of memory (or swap/disk space if that was used).

Reply:
Linux has a concept of "sparse files", so `fallocate` does not actually cause memory to be allocated. Thus, there is no concern about being OOM here with insanely large `fallocate` calls. See https://stackoverflow.com/questions/58571685/do-files-in-dev-shm-take-up-memory-when-grown-with-ftruncate-but-are-not-writte

Reply:
I've updated the tests to address the initial comment https://github.com/python-discord/snekbox/pull/113/commits/8028c08514a507e3de82a1260cbc79fe34095ccd, and I also bottomed out on the other thing I was confused on.